core/domain/exp_jobs_one_off.py

```diff
@@ -597,6 +602,62 @@ def reduce(status, values):
             yield (status, values)
 
 
+
+class ImagesAuditJob(jobs.BaseMapReduceOneOffJobManager):
+    """One-off job for checking whether all images in {{exp_id}}/assets/image/
+    are also present in exploration/{{exp_id}}/assets/image/.
+    """
+    @classmethod
+    def entity_classes_to_map_over(cls):
+        return [exp_models.ExplorationModel]
+
+    @staticmethod
+    def map(exp_model):
+        if exp_model.deleted:
+            return
+
+        if not constants.DEV_MODE:
+            exp_id = exp_model.id
+            fs_internal = fs_domain.AbstractFileSystem(
+                fs_domain.GcsFileSystem(exp_id))
+            fs_external = fs_domain.AbstractFileSystem(
+                fs_domain.GcsFileSystem('exploration/' + exp_id))
+            # We have to make sure we pass the dir name without starting or
+            # ending with '/'.
+            image_urls_internal = fs_internal.listdir('image')
+            image_urls_external = fs_external.listdir('image')
+
+            image_filenames_internal = [
+                GCS_IMAGE_ID_REGEX.match(url).group(3)
+                for url in image_urls_internal]
+
+            image_filenames_external = [
+                GCS_EXTERNAL_IMAGE_ID_REGEX.match(url).group(3)
+                for url in image_urls_external]
+
+            # Currently in editor.py and exp_services.py, all images are stored
+            # in exploration/{{exp_id}}/ parent folder. So, we only need to
+            # check if all images in the internal folder are there in the
+            # external folder.
```

Comment:
Easier: use sets instead. Then you can do set1.issubset(set2) -- or in fact just use set1.difference(set2) to find the missing files.

Reply:
Done